tfjs-backend-webgpu/src/backend_webgpu.ts

```diff
@@ -398,14 +412,119 @@ export class WebGPUBackend extends KernelBackend {
   // https://github.com/tensorflow/tfjs/issues/1595
   override readSync(dataId: object): BackendValues {
     const tensorData = this.tensorMap.get(dataId);
-    const {values} = tensorData;
+    const {values, complexTensorInfos} = tensorData;
 
-    if (values == null) {
-      throw new Error(
-          'WebGPU readSync is only available for CPU-resident tensors.');
+    if (values != null || tensorData.dtype === 'string') {
+      return values;
     }
 
-    return values;
+    if (tensorData.dtype === 'complex64') {
+      const realValues =
+          this.readSync(complexTensorInfos.real.dataId) as Float32Array;
+      const imagValues =
+          this.readSync(complexTensorInfos.imag.dataId) as Float32Array;
+      const complexVals = util.convertBackendValuesAndArrayBuffer(
+          backend_util.mergeRealAndImagArrays(realValues, imagValues).buffer,
+          'float32');
+      this.convertAndCacheOnCPU(dataId, complexVals);
```

Comment:
Can we remove the cache on CPU totally from now on?

Reply:
The right place is to remove https://github.com/tensorflow/tfjs/blob/master/tfjs-backend-webgpu/src/backend_webgpu.ts#L682-L686 and use `tensorData.values = null;` there.
Here, the `convertAndCacheOnCPU` may be still useful.

Reply:
Please also remove codes https://github.com/tensorflow/tfjs/blob/master/tfjs-backend-webgpu/src/base.ts#L27-L29.

Reply:
You can also change https://github.com/tensorflow/tfjs/blob/master/e2e/benchmarks/benchmark_util.js#L293 to add webgpu to the sync path so that you can use benchmarks' correctness tests for this PR.

Reply:
https://github.com/tensorflow/tfjs/blob/master/e2e/benchmarks/benchmark_util.js#L293 change is just to verify this PR's correctness (You can test it locally). Please make sure that even we switch to dataSync in benchmarks, the conformance test won't have any regressions.

Reply:
Jiajia, had tested locally, dataSync could work correctly on e2e cases.